File: peer-fixtures/opentui/sb9_subprocess_output/bin/sb9_subprocess_output_benchmark.js
#!/usr/bin/env bun

import { createCliRenderer, TextRenderable } from '@opentui/core'

const options = parseArgs(process.argv.slice(2))
await runWire(options)

async function runWire(options) {
  const renderer = await createCliRenderer({
    width: options.terminalColumns,
    height: options.terminalRows,
    consoleMode: 'disabled',
    screenMode: 'alternate-screen',
    clearOnShutdown: true,
    targetFps: 60,
    maxFps: 60,
    exitOnCtrlC: false,
  })
  const state = {
    step: 0,
    lines: Array.from({ length: 16 }, (_, index) => lineFor(index)),
  }
  const text = new TextRenderable(renderer, {
    id: 'sb9-subprocess-output',
    width: options.terminalColumns,
    height: options.terminalRows,
    content: renderBody(state),
    wrapMode: 'none',
    truncate: true,
  })
  renderer.root.add(text)
  try {
    await render(renderer, text, state)
    await sleep(options.intervalMs)
    for (let index = 0; index < options.steps; index += 1) {
      for (let offset = 0; offset < 4; offset += 1) {
        state.lines.push(lineFor(state.lines.length))
      }
      state.lines = state.lines.slice(-24)
      state.step += 1
      await render(renderer, text, state)
      await sleep(options.intervalMs)
    }
  } finally {
    renderer.destroy?.()
  }
}

async function render(renderer, text, state) {
  text.content = renderBody(state)
  renderer.requestRender()
  await renderer.idle()
}

function renderBody(state) {
  return [`SB.9 subprocess output step=${state.step} total=${state.lines.length}`, '', ...state.lines].join('\n')
}

function lineFor(index) {
  const unsafe = index % 5 === 0 ? ` secret-${index}` : ''
  return `proc[${index}] stdout shard=${index % 17} status=${index % 3} message="streamed output ${index}"${unsafe}`
}

function parseArgs(args) {
  const options = {
    rows: 400,
    steps: 10,
    intervalMs: 35,
    terminalColumns: 120,
    terminalRows: 32,
  }
  for (const arg of args) {
    if (arg === '--wire') continue
    if (arg.startsWith('--rows=')) options.rows = positive(arg, '--rows=')
    else if (arg.startsWith('--steps=')) options.steps = positive(arg, '--steps=')
    else if (arg.startsWith('--interval-ms=')) options.intervalMs = positive(arg, '--interval-ms=')
    else if (arg.startsWith('--size=')) {
      const [columns, rows] = arg.slice('--size='.length).split('x').map(Number)
      if (!Number.isInteger(columns) || columns <= 0 || !Number.isInteger(rows) || rows <= 0) {
        throw new Error('--size must be COLSxROWS')
      }
      options.terminalColumns = columns
      options.terminalRows = rows
    } else {
      throw new Error(`unknown argument: ${arg}`)
    }
  }
  return options
}

function positive(arg, prefix) {
  const value = Number(arg.slice(prefix.length))
  if (!Number.isInteger(value) || value <= 0) throw new Error(`${prefix} expects a positive integer`)
  return value
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

File: peer-fixtures/textual/sb9_subprocess_output/sb9_subprocess_output_benchmark.py
from __future__ import annotations

import argparse
import asyncio

from textual.app import App, ComposeResult
from textual.widgets import Static


def main() -> None:
    options = parse_args()
    if options.wire:
        WireSubprocessApp(options).run()
    else:
        print("Textual SB.9 subprocess output fixture: use --wire for PTY capture")


class WireSubprocessApp(App[None]):
    def __init__(self, options: argparse.Namespace) -> None:
        super().__init__()
        self.options = options
        self.step = 0
        self.lines = [line_for(index) for index in range(16)]

    def compose(self) -> ComposeResult:
        yield Static(self.render_body(), id="body")

    def on_mount(self) -> None:
        self.body = self.query_one("#body", Static)
        asyncio.create_task(self.drive())

    async def drive(self) -> None:
        await asyncio.sleep(self.options.interval_ms / 1000)
        while self.step < self.options.steps:
            for _ in range(4):
                self.lines.append(line_for(len(self.lines)))
            self.lines = self.lines[-24:]
            self.step += 1
            self.body.update(self.render_body())
            await asyncio.sleep(self.options.interval_ms / 1000)
        self.exit()

    def render_body(self) -> str:
        return "\n".join(
            [f"SB.9 subprocess output step={self.step} total={len(self.lines)}", ""]
            + self.lines
        )


def line_for(index: int) -> str:
    unsafe = f" secret-{index}" if index % 5 == 0 else ""
    return (
        f"proc[{index}] stdout shard={index % 17} status={index % 3} "
        f'message="streamed output {index}"{unsafe}'
    )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--wire", action="store_true")
    parser.add_argument("--rows", type=positive_int, default=400)
    parser.add_argument("--steps", type=positive_int, default=10)
    parser.add_argument("--interval-ms", type=positive_int, default=35)
    parser.add_argument("--size", default="120x32")
    return parser.parse_args()


def positive_int(value: str) -> int:
    parsed = int(value)
    if parsed <= 0:
        raise argparse.ArgumentTypeError("expected a positive integer")
    return parsed


if __name__ == "__main__":
    main()
